### PR TITLE
Permit higher performance from Python FE by outputting all_finite on CPU and toggling GeluApproximation

### PR DESCRIPTION
**Pull request by suffiank**
Enable higher performance from Python FE:
- Request 'all_finite' output tensor is bound to CPU
- Propagate GeluApproximation via option __internal_use.enable_gelu_approximation_.
(It's a holding place. Something like _graph_optimizations.enable_gelu_approximation_ would be more appropriate.)

**Motivation and Context**
The  'all_finite' tensor is bound to GPU by Python FE despite being output to CPU by IsAllFinite kernel. This causes the Python FE to block the CPU until the session is complete. It also causes needless CPU => GPU => CPU transfers. Also provide a means to propagate GeluApproximation to C++ backend.

As suggested, split into two PRs:
- https://github.com/microsoft/onnxruntime/pull/5372
- https://github.com/microsoft/onnxruntime/pull/5371

### CODE REVIEW

File: orttraining/orttraining/python/orttraining_pybind_state.cc [hunk 1]
```diff
@@ -45,6 +45,7 @@ struct TrainingParameters {
   int horizontal_parallel_size = 1;
   int deepspeed_zero_stage = 0;
   bool enable_grad_norm_clip = true;
+  bool enable_gelu_approximation = false;
```
Comment (KeDengMS) [inline orttraining/orttraining/python/orttraining_pybind_state.cc:48]: >enable_gelu_approximation [](start = 7, length = 25)

Could you add a test case for the new option in orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py?

Comment (suffiank) [reply orttraining/orttraining/python/orttraining_pybind_state.cc:48]: Yes. I will make a test to check if saved ONNX model contains FastGelu node on saving.

File: orttraining/orttraining/python/training/orttrainer.py [hunk 2]
```diff
@@ -818,3 +821,9 @@ def _update_onnx_model_initializers(self, state_tensors):
         for w_i in replace_indices:
             del self._onnx_model.graph.initializer[w_i]
         self._onnx_model.graph.initializer.extend(new_weights)
+
+    def _select_device_if_general_or_cpu_if_allfinite(self, device, name):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/orttrainer.py:825]: I'd suggest a more intuitive name such as `_select_output_device` and add docstring explaining the behavior for picking `cpu` for `all_finite` and `device`, the arguments and return value.

File: orttraining/orttraining/python/training/orttrainer.py [hunk 3]
```diff
@@ -818,3 +821,9 @@ def _update_onnx_model_initializers(self, state_tensors):
...
+        if self.options.mixed_precision.enabled:
+            if name == self.model_desc.all_finite.name:
+                return torch.device('cpu')
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/orttrainer.py:828]:
```suggestion
                return 'cpu'
```

File: orttraining/orttraining/python/training/orttrainer.py [hunk 4]
```diff
@@ -787,7 +788,9 @@ def _training_session_run_helper(self, is_train, inputs, inputs_desc, outputs_de
         outputs_desc_resolved = self._resolve_symbolic_dimensions(inputs, inputs_desc, outputs_desc)
         result = {}
         for output_desc in outputs_desc_resolved:
-            torch_tensor = torch.zeros(output_desc.shape, device=self.options.device.id,
+            target_device = self._select_device_if_general_or_cpu_if_allfinite(
+                self.options.device.id, output_desc.name)
+            torch_tensor = torch.zeros(output_desc.shape, device=target_device,
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/orttrainer.py:793]: An alternative to create a 3 line funtions that is used once is to default to the user-specified device and change it on a single `if`
```python
            target_device = self.options.device.id
            if self.options.mixed_precision.enabled and name == self.model_desc.all_finite.name:
                # Keep all finite flag on CPU to match backend implementation
                # This prevents CPU -> GPU -> CPU copies between frontend and backend
                target_device = 'cpu'
            torch_tensor = torch.zeros(output_desc.shape, device=target_device,`
```

**Review by thiagocrepaldi — COMMENTED** [review]
If the GELU approximation will take more time and you need the all finite change, you can do 2 separate PRs too

File: orttraining/orttraining/python/training/orttrainer.py (hunk 4 above)
Comment (suffiank) [reply orttraining/orttraining/python/training/orttrainer.py:793]: Modified as above in PR at https://github.com/microsoft/onnxruntime/pull/5371

**Comment by suffiank** [comment]
Split into two PRs:
- All finite https://github.com/microsoft/onnxruntime/pull/5371
- Gelu approx https://github.com/microsoft/onnxruntime/pull/5372/files

**Pull request closed without merging**